pkgs/applications/editors/neovim/wrapper.nix

```diff
@@ -55,46 +55,95 @@ let
+  # add to nvim's 'embedded rc' this:
+  #
+  #    let g:<key>_host_prog=$out/bin/nvim-<key>
+  #
+  # Or this:
+  #
+  #    let g:loaded_${prog}_provider=1
+  #
+  # While the later tells nvim that this provider is not available
+  #
+  hostprog_check_table = {
+    node = withNodeJs;
+    python = withPython;
+    python3 = withPython3;
+    ruby = withRuby;
+  };
+  # For every "host" program neovim supports, we map according to the attrset
+  # `hostprog_check_table` what flags to add to wrapper
+  nvimFlags = lib.attrsets.mapAttrsToList (
+      prog:
+      withProg:
+      # --cmd is needed even if we disable this ${prog} provider or not
+      ["--cmd" (lib.escapeShellArg
+        (if withProg then
+          # All host programs end up in the same path
+          "let g:${prog}_host_prog='${placeholder "out"}/bin/nvim-${prog}'"
+        else
+          # Tells nvim to practically not load this provider
+          "let g:loaded_${prog}_provider=1"
+        ))
+      ]
+  ) hostprog_check_table;
+  ## Here we calculate all of the arguments to `makeWrapper`
+  # We start with the executable itself
+  makeWrapperArgs = [ "${neovim}/bin/nvim" "${placeholder "out"}/bin/nvim" "--argv0" "$0" ]
+    ++ (lib.lists.flatten (lib.lists.forEach nvimFlags (
+      flag:
+      # Already shell-escaped
+      [ "--add-flags" (lib.concatStringsSep " " flag) ]
+    )))
+    ++ lib.optionals withRuby [
+      "--set" "GEM_HOME" "${rubyEnv}/${rubyEnv.ruby.gemPath}"
+    ]
+    ++ lib.optionals (binPath != "") [
+      "--suffix" "PATH" ":" binPath
+    ]
+    # this relies on a patched neovim, see
+    # https://github.com/neovim/neovim/issues/9413
+    ++ lib.optionals (configure != {}) [
+      "--set" "NVIM_SYSTEM_RPLUGIN_MANIFEST" "$out/rplugin.vim"
+      "--add-flags" "-u ${vimUtils.vimrcFile configure}"
+    ]
+  ;
+  cleanNvimFlags = lib.concatStringsSep " " (lib.lists.flatten nvimFlags);
   in
-  stdenv.mkDerivation {
+  symlinkJoin {
       name = "neovim-${stdenv.lib.getVersion neovim}";
-      buildCommand = let bin="${neovim}/bin/nvim"; in ''
-        if [ ! -x "${bin}" ]
-        then
-            echo "cannot find executable file \`${bin}'"
-            exit 1
-        fi
-
-        makeWrapper "$(readlink -v --canonicalize-existing "${bin}")" \
-          "$out/bin/nvim" --add-flags " \
-        --cmd \"${if withNodeJs then "let g:node_host_prog='${nodePackages.neovim}/bin/neovim-node-host'" else "let g:loaded_node_provider=1"}\" \
-        --cmd \"${if withPython then "let g:python_host_prog='$out/bin/nvim-python'" else "let g:loaded_python_provider = 1"}\" \
-        --cmd \"${if withPython3 then "let g:python3_host_prog='$out/bin/nvim-python3'" else "let g:loaded_python3_provider = 1"}\" \
-        --cmd \"${if withRuby then "let g:ruby_host_prog='$out/bin/nvim-ruby'" else "let g:loaded_ruby_provider=1"}\" " \
-        --suffix PATH : ${binPath} \
-        ${optionalString withRuby '' --set GEM_HOME ${rubyEnv}/${rubyEnv.ruby.gemPath}'' }
```

Comment:
Hm it's closer to the original now, but there's still some differences. For one, `GEM_HOME` isn't set anymore when generating the `rplugin.vim`, same for `PATH`. The current code can really only deal with command line arguments though.

What could be done instead is to compute a shared set of `makeWrapper` args as here first (which will not include the `rplugin.vim` ones), wrap the binary with those, call it for determining `rplugin.vim`, then call `makeWrapper` on the original unwrapped binary once again, but this time including the `rplugin.vim` options